Refactor UFS-DA ATM var and ens components to use python g-w

RussTreadon-NOAA opened the pull request:
**Description**
g-w is moving toward a python based approach.   The UFS-DA Aerosol component uses the python g-w approach.  This PR  replaces the non-python based UFS-DA ATM var and ens components with python based equivalents.

Fixes #1313 

**Type of change**

- [x] Update feature (non-breaking change which refactors existing functionality)


**How Has This Been Tested?**

- [ ] Clone and Build tests on Hera and Orion
- [x] Cycled test on Hera

  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

RussTreadon-NOAA committed:
update UFS-DA atm var components to python g-w (#1313)

RussTreadon-NOAA committed:
fix pynorms errors (#1313)

RussTreadon-NOAA committed:
add config files for UFS-DA atm var python g-w (#1313)

RussTreadon-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/refactor_jediatm

RussTreadon-NOAA committed:
update UFS-DA atm ens components to python g-w (#1313)

RussTreadon-NOAA committed:
remove UFS-DA atm var and ens components using non-python g-w (#1313)

RussTreadon-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/refactor_jediatm

RussTreadon-NOAA committed:
address ShellCheck errors in config files (#1313)

WalterKolczynski-NOAA commented on jobs/JGLOBAL_ATMENS_ANALYSIS_FINALIZE:
```diff
@@ -24,24 +26,21 @@ export OPREFIX="${CDUMP}.t${cyc}z."
-# Add UFSDA to PYTHONPATH
-export PYTHONPATH=${HOMEgfs}/sorc/gdas.cd/ush/:${PYTHONPATH}
+export COMIN_GES="${ROTDIR}/${GDUMP}.${gPDY}/${gcyc}/chem"
+export COMIN_GES_ENS="${ROTDIR}/enkf${GDUMP}.${gPDY}/${gcyc}/chem"
```
All of these scripts are now using `chem`. Shouldn't these still be `atmos`? That's what it was before and the script name indicates ATM analysis, not aerosol.

RussTreadon-NOAA replied:
`chem` has been replaced with `atmos` in jobs.  `chem` entries in `ush/python/pygfs/task/atm_analysis.py` and `ush/python/pygfs/task/atmens_analysis.py` have been removed.   Done at 6ae4c3a8.

WalterKolczynski-NOAA commented on parm/config/config.atmanl:
```diff
@@ -19,7 +19,13 @@ export R2D2_BC_DUMP='oper_gdas'
 
-echo "END: config.atmanal"
+export JEDIVAREXE=${HOMEgfs}/exec/fv3jedi_var.x
+export CRTM_VER="2.3.0"
```
The crtm module is already providing the version as `$crtm_VERSION`. That should be used instead of defining it here, which could lead to a mismatch.

RussTreadon-NOAA replied:
`config.atmanl` followed what is in `config.aeroanl`.   Removed `CRTM_VER` from `config.atmanl`, `config.atmensanl`, and `config.aeroanl`.   Doing so required replacing `CRTM_VER` with `crtm_VERSION` from `parm/parm_gdas/aero_crtm_coeff.yaml` amd `parm/parm_gdas/atm_crtm_coeff.yaml`.   Done at 6ae4c3a8.

WalterKolczynski-NOAA commented on parm/config/config.atmensanl:
```diff
@@ -0,0 +1,31 @@
+export io_layout_y=1
+
+export JEDIVAREXE=${HOMEgfs}/exec/fv3jedi_var.x
+export CRTM_VER="2.3.0"
```
As before.

RussTreadon-NOAA replied:
Resolved as noted above.  Done at 6ae4c3a8.

RussTreadon-NOAA replied:
@WalterKolczynski-NOAA 

Unfortunately this is not resolved.   When we run UFS DA, workflow jobs execute `ush/load_ufsda_modules.sh`.   This loads `sorc/gdas.cd/modulefiles/GDAS/$machine.lua` where `machine` is `hera` or `orion` depending on the platform on which you are running.     The following crtm modules are available on Hera and Orion.
```
     Versions:
        crtm/v2.3-jedi.4
        crtm/v2.4_jedi
        crtm/2.3.0
        crtm/2.4.0
```
Of these, only the Orion `crtm/2.4.0` defines `crtm_VERSION`.    None of the Hera crtm modules define `crtm_VERSION`.   We can not assume `crtm_VERSION` is defined when running UFS DA.    

The Hera and Orion modulefiles in `gdas.cd/modulefiles/GDAS` do not load a crtm module.  Instead, `CRTM_FIX` is defined.    `CRTM_VER` is specified in UFS DA config files to identify the specific version of `CRTM_FIX` to use.    Currently the UFS DA `CRTM_FIX` only contains coefficients for `2.3.0` and `2.3.0_jedi`.    

For the time being we need to set `CRTM_VER` in UFS DA config files.     Hopefully future updates to UFS DA crtm modules will define `crtm_VERSION` on all supported machines.

Tagging @CoryMartin-NOAA for awareness.

RussTreadon-NOAA replied:
Since I already updated `aero_crtm_coeff.yaml` and `atm_crtm_coeff.yaml` to `crtm_VERSION`, I added `export crtm_VERSION=2.3.0` to `config.atmanl` as a test.   `gdasatmanlinit` was submitted.   It successfully ran to completion with the correct CRTM coefficients staged.   This isn't an ideal solution, but it works.

RussTreadon-NOAA replied:
Added `crtm_VERSION` to `config.atmanl` and `config.atmensanl` at c207ff66.

WalterKolczynski-NOAA commented on parm/parm_gdas/atm_crtm_coeff.yaml:
```diff
@@ -0,0 +1,350 @@
+- !ENV ${DATA}/crtm/
+copy:
+# Emissivity files
+- - !ENV ${FV3JEDI_FIX}/crtm/${CRTM_VER}/NPOESS.VISice.EmisCoeff.bin
```
See above about using the CRTM version defined by the module (`$crtm_VERSION`) instead of this different variable.

RussTreadon-NOAA replied:
Resolved as noted above.  Done at 6ae4c3a8.

WalterKolczynski-NOAA commented on parm/parm_gdas/atmanl_inc_vars.yaml:
```diff
@@ -0,0 +1 @@
+['dust1', 'dust2', 'dust3', 'dust4', 'dust5', 'seas1', 'seas2', 'seas3', 'seas4', 'so4', 'oc1', 'oc2', 'bc1', 'bc2']
```
Cory just made a change in PR #1349 that changed this in the aerosol DA to be `incvars: [<var list>]`. Does this need to be similarly changed?

RussTreadon-NOAA replied:
Revise `parm/parm_gdas/atmanl_inc_vars.yaml` to be consistent with PR #1349.  Also update variables to be those for ATM, not Aerosol, DA.   Done at 6ae4c3a8.

WalterKolczynski-NOAA commented on ush/python/pygfs/task/analysis.py:
```diff
@@ -22,14 +22,21 @@ class Analysis(Task):
         FileHandler(obs_dict).sync()
 
+    def initialize_bias(self) -> None:
+        super().initialize()
```
`Task` doesn't have an `initialize_bias()` method.

RussTreadon-NOAA replied:
`initialize_bias` originated as a copy of `initialize`.   While all analyses needs to stage observations, not all analyses need to stage bias correction files.   `initialize_bias` was added to provide this functionality.   

I'm not proficient in python.   I thought your comment meant that the `super().initialize()` line in `initialize_bias` should read `super().initialize_bias()`.  I tried this.  It doesn't work.

The next thing I tried was to move scripting from `initialize_bias` into `initialize` so that we have
```
    def initialize(self) -> None:
        super().initialize()
        # all analyses need to stage observations
        obs_dict = self.get_obs_dict()
        FileHandler(obs_dict).sync()

        # some analyses need to bias corrections
        bias_dict = self.get_bias_dict()
        FileHandler(bias_dict).sync()
```
The block 
```
        # stage bias correction files
        super().initialize_bias()
```
was removed from `atm_analysis.py` leaving only
```
        # stage observations and bias corrections
        super().initialize()
```
`gdasatmanlinit` was submitted.  It sucessfully ran to completion.   Bias correction files were correctly staged.   Is this an acceptable resolution to your comment @WalterKolczynski-NOAA ?

RussTreadon-NOAA replied:
The above change was committed at c207ff66.

WalterKolczynski-NOAA replied:
That's correct. The error is what I expected. The parent class (`Task`) doesn't have the method `initialize_bias()`. `super.initialize_bias()` attempts to call the method in the parent class. Therefore the error.

This solution is probably the best one.

WalterKolczynski-NOAA commented on ush/python/pygfs/task/atmens_analysis.py:
```diff
@@ -0,0 +1,295 @@
+        FileHandler(jedi_fix_list).sync()
+
+        # stage backgrounds
+        FileHandler(self.get_bkg_dict(AttrDict(self.task_config, **self.task_config))).sync()
```
A lot of this looks the same as the non-ensemble class. Can more of it be abstracted up into the parent class?

RussTreadon-NOAA replied:
Attempt to implement this suggestion made at c207ff66.

WalterKolczynski-NOAA commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -1102,57 +1086,52 @@ def eupd(self):
-        dmpdir = self._base["DMPDIR"]
         do_gfs_enkf = True if self.app_config.do_hybvar and 'gfs' in self.app_config.eupd_cdumps else False
+        cycledef = 'gdas'
+        if self.cdump in ['gfs'] and do_gfs_enkf:
+            cycledef = 'gfs'
+            if gfs_cyc != 4:
+                cycledef = 'gdas'
```
Not sure what you are trying to do here. If this needs to run every cycle even if the free forecast is not, you can just run on the `gdas` `cycledef` regardless of `cdump`.

RussTreadon-NOAA replied:
This section replicates, in part, what was in the removed `atmensanalprep` section.
```
        cycledef = self.cdump
        if self.cdump in ['gfs'] and do_gfs_enkf and gfs_cyc != 4:
            cycledef = 'gdas'
```
I assumed this was something we need to retain due to the option to now run a partial enkf cycle in the gfs as well as the full enkf cycle in the gdas.  I was not part of this change and so I don't know the details.  

The new `atmensanlinit` step needs to run after the prep step.  Its workflow setup is similar to the existing `eobs`.   I updated this section of `workflow_tasks.py` to minic `eobs`.  Done at 3a3d6288.

WalterKolczynski-NOAA replied:
Yeah, what was originally there was unnecessary too. If it needs to run every cycle regardless of how often the full forecast runs, the `cycledef` should be `gdas`, regardless of the job's actual `cdump` (or `gdas_half,gdas` if it also needs to run in the half-cycle).

WalterKolczynski-NOAA commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -1102,57 +1086,52 @@ def eupd(self):
         return task
 
-    def atmensanalrun(self):
+    def atmensanlrun(self):
```
Do these other tasks in the atmensanl job sequence need to run in the same `cycledefs` as the `init`?

RussTreadon-NOAA replied:
Yes, `atmensanlinit`, `atmensanlrun`, and `atmensanlfinal` are an initialize, run, finalize group which need to sequentially run in the same cycle.    Not sure this answers your question.

WalterKolczynski-NOAA reviewed (changes requested):
I am very confused by this PR. It's described as updating the ATM DA. The filenames indicate it is for ATM DA. But in many places it looks like the functionality is being changed to be aerosol DA (output directory is changed to `chem`, addition of aerosol trace list into yaml file). But there is already an aerosol DA for UFSDA. Can you help me understand what is going on?

Also seems like there is a bunch of repeated code between the two atmanl classes. I think a lot more of it can be abstracted up into the parent class.

RussTreadon-NOAA commented:
Thank you @WalterKolczynski-NOAA for looking over this PR, asking great questions, and making good suggestions.  This is a big PR.  It began as a copy UFS DA Aerosol.   I'll replace the occurrences of `chem` with `atmos`.    There are things in the original UFS DA Aerosol and python g-w which I'm not sure are entirely correct.    With regards to abstraction, I don't disagree but will need help to achieve this.

RussTreadon-NOAA committed:
replace chem with atmos, CRTM_VER with crtm_VERSION, update atmanl_inc variables (#1313)

RussTreadon-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/refactor_jediatm

RussTreadon-NOAA commented:
@WalterKolczynski-NOAA and @aerorahul
While the automated shellcheck flags many lines in `parm/config/config.resources`, I do not get these errors when I execute `/home/Rahul.Mahajan/bin/shellcheck` or `/scratch2/NCEPDEV/ensemble/save/Walter.Kolczynski/shellcheck-v0.8.0/shellcheck` on `config.resources`

```
Hera(hfe07):/scratch1/NCEPDEV/da/Russ.Treadon/git/global-workflow/test2/parm/config$ /home/Rahul.Mahajan/bin/shellcheck config.resources
Hera(hfe07):/scratch1/NCEPDEV/da/Russ.Treadon/git/global-workflow/test2/parm/config$
Hera(hfe07):/scratch1/NCEPDEV/da/Russ.Treadon/git/global-workflow/test2/parm/config$ /scratch2/NCEPDEV/ensemble/save/Walter.Kolczynski/shellcheck-v0.8.0/shellcheck config.resources
Hera(hfe07):/scratch1/NCEPDEV/da/Russ.Treadon/git/global-workflow/test2/parm/config$
```

Furthermore, the automated shellcheck messges for `config.resources` are confusing.   For example, the automated check reports

```
Error: SHELLCHECK_WARNING:
./parm/config/config.resources:7:4: note[SC2292]: Prefer [[ ]] over [ ] for tests in Bash/Ksh.

Error: SHELLCHECK_WARNING:
./parm/config/config.resources:52:4: note[SC2292]: Prefer [[ ]] over [ ] for tests in Bash/Ksh.
```

Lines 7 and 52 of `config.resources` use double brackets.
```
if [[ $# -ne 1 ]]; then
...
if [[ "${step}" = "prep" ]]; then
```

Am I misunderstanding the [lint-shell report](https://github.com/NOAA-EMC/global-workflow/actions/runs/4314976035/jobs/7528718636)?

Is it only line 9 of `config.atmanlinit` which needs to be fixed?   That is, 
```
 # Get task specific resources
. ${EXPDIR}/config.resources atmanlinit
```
needs to be rewritten as
```
 # Get task specific resources
. "${EXPDIR}"/config.resources atmanlinit
```

RussTreadon-NOAA committed:
address shellnorms error in config.atm files (#1313)

RussTreadon-NOAA commented:
@WalterKolczynski-NOAA & @aerorahul , please disregard my previous comment.  I fixed shellcheck errors in `config.atmanlinit` at [1536262](https://github.com/NOAA-EMC/global-workflow/pull/1354/commits/153626254371194c5a14fc913924feb98b4dad7b).   `config.resources` was not flagged by the rerun of shellcheck.

RussTreadon-NOAA committed:
simplify workflow setup for atmensanlinit (#1313)

RussTreadon-NOAA commented on parm/config/config.atmanlinit:
```diff
@@ -0,0 +1,10 @@
+echo "BEGIN: config.atmanlinit"
+
+# Get task specific resources
+. ${EXPDIR}/config.resources atmanlinit
```
Fixed at [1536262](https://github.com/NOAA-EMC/global-workflow/pull/1354/commits/153626254371194c5a14fc913924feb98b4dad7b).

RussTreadon-NOAA committed:
add crtm_VERSION to atmanl and atmensanl config; move common functionality from analysis task class to parent class (#1313)

RussTreadon-NOAA commented:
@WalterKolczynski-NOAA , I worked through your comments an attempted to address them all.   If I misunderstood any of your comments, my apologies.   As noted above I am not proficient with python.  I'm still learning concepts, syntax, and lingo.  Correct me as needed and I'll try again.   Also let me know if I didn't adequately answer any of your questions.   Thanks.

RussTreadon-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/refactor_jediatm

KateFriedman-NOAA commented:
@RussTreadon-NOAA Checking if the `export crtm_VERSION="2.3.0"` settings conflict with similar crtm version settings in the rest of the system. Other crtm version settings currently:

1. We moved to `crtm/2.4.0` recently in `develop` and currently have a hack in `build_gsi_enkf.sh` to force the GSI to build with `2.4.0` while waiting for a new GSI hash that uses `2.4.0` (will remove hack then): https://github.com/NOAA-EMC/global-workflow/blob/develop/sorc/build_gsi_enkf.sh#L23
2. These `crtm_VERSION` variables conflict with the naming convention for other version variables (as we introduced with GFSv16: lowercase `"thing_ver"`). We'll be introducing version files in the near future (once GSI moves away from intel 2018, current blocker). [I have set my version file feature branch](https://github.com/KateFriedman-NOAA/global-workflow/tree/feature/version-files) to use `crtm_ver` currently (like ops). If the decision is to go to `crtm_VERSION` then I need to know. Is `crtm_VERSION` intended to be used like `crtm_ver` or to replace `crtm_ver`?

Thanks!

RussTreadon-NOAA commented:
@KateFriedman-NOAA , I don't know the answer to your question about `crtm_VERSION` and `crtm_ver`.   I did not know about `crtm_VERSION` until @WalterKolczynski-NOAA  mentioned it.  The library team should be able to answer your good question.

WalterKolczynski-NOAA commented:
> @RussTreadon-NOAA Checking if the `export crtm_VERSION="2.3.0"` settings conflict with similar crtm version settings in the rest of the system. Other crtm version settings currently:
> 
> 1. We moved to `crtm/2.4.0` recently in `develop` and currently have a hack in `build_gsi_enkf.sh` to force the GSI to build with `2.4.0` while waiting for a new GSI hash that uses `2.4.0` (will remove hack then): https://github.com/NOAA-EMC/global-workflow/blob/develop/sorc/build_gsi_enkf.sh#L23
> 2. These `crtm_VERSION` variables conflict with the naming convention for other version variables (as we introduced with GFSv16: lowercase `"thing_ver"`). We'll be introducing version files in the near future (once GSI moves away from intel 2018, current blocker). [I have set my version file feature branch](https://github.com/KateFriedman-NOAA/global-workflow/tree/feature/version-files) to use `crtm_ver` currently (like ops). If the decision is to go to `crtm_VERSION` then I need to know. Is `crtm_VERSION` intended to be used like `crtm_ver` or to replace `crtm_ver`?
> 
> Thanks!

Just to make sure we are all the same page:
1. `crtm_ver` is used by us to chooses which module to load
2. `crtm_VERSION` is then defined by that module (at least in hpc-stack on Orion)

WalterKolczynski-NOAA commented on parm/config/config.atmanlfinal:
```diff
@@ -6,5 +6,5 @@
 
 # Get task specific resources
-. "${EXPDIR}/config.resources atmanlfinal"
+. "${EXPDIR}"/config.resources atmanlfinal
```
We prefer canonical quotations (full path). This code failed before because you put quotes around two different arguments. Try this:

```suggestion
. "${EXPDIR}/config.resources" atmanlfinal
```

RussTreadon-NOAA replied:
Suggestion committed at 96b89d80.

WalterKolczynski-NOAA commented on parm/config/config.atmanlinit:
```diff
@@ -6,5 +6,5 @@
 
 # Get task specific resources
-. "${EXPDIR}/config.resources atmanlinit"
+. "${EXPDIR}"/config.resources atmanlinit
```
```suggestion
. "${EXPDIR}/config.resources" atmanlinit
```

RussTreadon-NOAA replied:
Suggestion committed at 96b89d80.

WalterKolczynski-NOAA commented on parm/config/config.atmanlrun:
```diff
@@ -6,6 +6,6 @@
 
 # Get task specific resources
-. "${EXPDIR}/config.resources atmanlrun"
+. "${EXPDIR}"/config.resources atmanlrun
```
```suggestion
. "${EXPDIR}/config.resources" atmanlrun
```

RussTreadon-NOAA replied:
Suggestion committed at 96b89d80.

WalterKolczynski-NOAA commented on parm/config/config.atmensanlfinal:
```diff
@@ -6,5 +6,5 @@
 
 # Get task specific resources
-. "${EXPDIR}/config.resources atmanlfinal"
+. "${EXPDIR}"/config.resources atmanlfinal
```
```suggestion
. "${EXPDIR}/config.resources" atmanlfinal
```

RussTreadon-NOAA replied:
Suggestion committed at 96b89d80.

WalterKolczynski-NOAA commented on parm/config/config.atmensanlinit:
```diff
@@ -6,5 +6,5 @@
 
 # Get task specific resources
-. "${EXPDIR}/config.resources atmanlinit"
+. "${EXPDIR}"/config.resources atmanlinit
```
```suggestion
. "${EXPDIR}/config.resources" atmanlinit
```

RussTreadon-NOAA replied:
Suggestion committed at 96b89d80.

WalterKolczynski-NOAA commented on parm/config/config.atmensanlrun:
```diff
@@ -6,6 +6,6 @@
 
 # Get task specific resources
-. "${EXPDIR}/config.resources atmanlrun"
+. "${EXPDIR}"/config.resources atmanlrun
```
```suggestion
. "${EXPDIR}/config.resources" atmanlrun
```

RussTreadon-NOAA replied:
Suggestion committed at 96b89d80.

WalterKolczynski-NOAA commented on workflow/rocoto/workflow_tasks.py:
```diff
@@ -1086,25 +1086,15 @@ def eupd(self):
         resources = self.get_resource('atmensanlinit')
-        task = create_wf_task('atmensanlinit', resources, cdump=self.cdump, envar=self.envars, dependency=dependencies,
-                              cycledef=cycledef)
+        task = create_wf_task('atmensanlinit', resources, cdump=self.cdump, envar=self.envars, dependency=dependencies)
```
See my other comment about the `cycledef`. I think you may still need it here, but just set to `gdas` for all cdumps.

RussTreadon-NOAA replied:
Suggestion committed at 96b89d80.

WalterKolczynski-NOAA reviewed (changes requested):
Much better. Glad I wasn't missing something and it was supposed to be for atm. Cycledefs may need more changes, plus a minor syntaxial preference. Please be sure to test it thoroughly to make sure it is doing what you expect.

KateFriedman-NOAA commented:
> Just to make sure we are all the same page:
> 
> 1. `crtm_ver` is used by us to chooses which module to load
> 2. `crtm_VERSION` is then defined by that module (at least in hpc-stack on Orion)

Thanks @WalterKolczynski-NOAA , I forgot about that internal variable in the module. 

Ok, so I see the internal module variable is being used to define the path to the crtm files:
```
!ENV ${FV3JEDI_FIX}/crtm/${crtm_VERSION}/v.viirs-m_j1.SpcCoeff.bin
```
...and I see some hardcoding in configs (`export crtm_VERSION="2.3.0"`) to use v2.3.0. @RussTreadon-NOAA does UFS-DA still need to use `crtm/2.3.0` and these hardcoded settings reset the version after the workflow loaded the `crtm/2.4.0` module? I see they were added in this commit: https://github.com/NOAA-EMC/global-workflow/pull/1354/commits/c207ff6636c116907c6dc89f7f78d78c873d9b2b Just wanna understand the usage and stay aware of any component adjusting the version. Thanks! :)

RussTreadon-NOAA commented:
@KateFriedman-NOAA , UFS DA unified forward operator (UFO) validation is being done with crtm/2.3.0.   So for the time being we reference crtm/2.3.0 in the UFS DA atm and ens components.   Once UFO validation is complete UFS DA may skip crtm/2.4.0 and jump to using crtm/3.0.0.

KateFriedman-NOAA commented:
> UFS DA unified forward operator (UFO) validation is being done with crtm/2.3.0. So for the time being we reference crtm/2.3.0 in the UFS DA atm and ens components. Once UFO validation is complete UFS DA may skip crtm/2.4.0 and jump to using crtm/3.0.0.

@RussTreadon-NOAA Understood, thanks for explaining that and the plan! :)

KateFriedman-NOAA commented:
@RussTreadon-NOAA Oh, also, definitely let us know if/when you plan to move to a version after 2.4.0 (or whatever the system as a whole is using at the time) so we can coordinate potentially moving other components to the same version. Thanks!

RussTreadon-NOAA committed:
incorporate reviewer suggestions, clean up UFS-DA var and ens config files, correct errors in config.atmensanl, remove unnecessary line in analysis.py (#1313)

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/refactor_jediatm

RussTreadon-NOAA commented:
@WalterKolczynski-NOAA , I incorporated your suggestions.   I appreciate the input.  While testing I found and corrected problems with `config.atmensanl` and `analysis.py`.    

`config.resources` in `RussTreadon-NOAA:feature/refactor_jediatm` at 96b89d80 had conficts with the authoritative `develop` at 47afc785.   The conflicts were manually resolved and `RussTreadon-NOAA:feature/refactor_jediatm` brougt up to date with `develop` at 4e9932ba.

Jobs `gdasatmanl[init, run, final]` and `enkfgdasatmanl[init, run, final]` were rewound and rebooted.   Each job successfully ran to completion.   We are not yet testing UFS-DA ATM var or ens in the gfs cycle.   The first step is to get the gdas cycle working.

RussTreadon-NOAA commented:
Thank you @WalterKolczynski-NOAA for looking over the latest updates and finding them to be OK.   I appreciate your guidance in the review process.

EIB has requested that the changes in this PR be split into three PRs.  This PR will not be merged into `develop` as is.   Instead this PR will be refactored as follows:
1. Create a new forked branch and a new PR for changes outside UFS-DA ATM var and ens components.
2. Create a new forked branch and a new PR once the PR for 1 is merged into `develop`.   The new branch contains UFS-DA ATM var changes.
3. Create a new forked branch and a new PR once the PR for 2 is merged into `develop`.   The new branch contains UFS-DA ATM ens changes. 

Some files contain changes to multiple components.   Thus, some changes related to later PRs may be present in earlier PRs.

RussTreadon-NOAA commented:
Close as per EIB request.
